### ISSUE TITLE
[Proposal] Static property 'table' in Model class 

### ISSUE THREAD

**Issue by allok** [laravel/ideas#1305]
Now in the 'Model' class we have:   
```
     /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table;
```
It will be good to have the static property 'table' which gives a number of advantages:
1) Easy to use in queries with join (where you need to explicitly indicate table name).
2) Easy to use in plain DB queries.
3) Easy to use in every code place where need it.

**Comment by sisve** [comment]
Are there no disadvantages?

**Comment by allok** [comment]
@sisve maybe it is not support simple override for each object. But we have getTable method()

**Comment by AkenRoberts** [comment]
@fletch3555 They're proposing making it a public static property.

👍 on making it static, along with a lot of the other properties. It would be a large breaking change, though.

**Comment by drbyte** [comment]
Question about this, related to using a non-empty `config(db.<driver>.prefix)`...

When you use the `$table` property directly in either plain DB queries or join strings, does the query-builder or grammar parser intelligently parse those strings for all mentions of your self-injected table-name and manually add the `config(db.<driver>.prefix)` automagically every time? perfectly?

Or are you just always using a blank `prefix` setting, so never run into this?

Does using `getTable()` take care of adding the `prefix` in, or is it equally ignorant of non-blank prefixes?

(I've never had to use non-empty prefixes in my Laravel apps, so haven't run into this, but in other apps it's been a huge source of pain. Hence the question.)

**Comment by AkenRoberts** [comment]
You can overwrite a static property just like you would the current instance property.

**Comment by sisve** [comment]
@AkenRoberts Wouldn't that change all instances, and not just the current one?

    $user1 = new User();
    $user2 = new User();
    $user2->setTable('another_table');

**Comment by allok** [comment]
@sisve @AkenRoberts 
We can define two properties in Model like:
```
    /**
     * The table associated with the model.
     *
     * @var string
     */
    public static $table;

    /**
     * The table associated with the object of model.
     *
     * @var string
     */
    protected $currentTable;

    /**
     * Get the table associated with the model.
     *
     * @return string
     */
    public function getCurrentTable()
    {
        if (! isset($this->currentTable)) {
            $this->currentTable = self::getTable();
        }

        return $this->currentTable;
    }

    public static function getTable()
    {
            return self::$table ?? str_replace(
                '\\', '', Str::snake(Str::plural(class_basename($this)))
            );
    }
```

**Comment by sisve** [comment]
Couldn't you just call `User::getTable()` today? That would give you the default value.

**Comment by AkenRoberts** [comment]
> @AkenRoberts Wouldn't that change all instances, and not just the current one?

Yeah, with the static property, all instances of User would have the same table, unless you overwrote it at the instance level which you could still do.

**Comment by robjbrain** [comment]
Came here to suggest the same thing. If you had `$table` as static you could potentially even make `qualifyColumn` as static as well.

That way instead of doing

```
User::join('posts', 'posts.user_id', '=', 'users.id')->get();
```

You could do:

```
User::join(Post::getTable(), Post::qualifyColumn('user_id'), '=', User::qualifyColumn('id'))->get();
```

Then if you ever change the "posts" table to "articles" you only change it in one place, in the model.

Right now if you wanted to do the same you'd have to get an instance of the model each time

```
User::join((new Post)->getTable(), (new Post)->qualifyColumn('user_id'), '=', (new User)->qualifyColumn('id'))->get();
```

This would affect the ability to use:
`setKeyName` 
`getQualifiedKeyName`
`setTable`

Perhaps this could be mitigated with a local variable as suggested above?
